include/engine/guidance/segment_list.hpp

```diff
@@ -0,0 +1,304 @@
+}
+
+template <typename DataFacadeT>
+void SegmentList<DataFacadeT>::Finalize(const bool extract_alternative,
+                                        const InternalRouteResult &raw_route,
+                                        const unsigned zoom_level,
+                                        const bool allow_simplification)
+{
+    if (segments.empty())
+        return;
+
+    segments[0].length = 0.f;
+    for (const auto i : osrm::irange<std::size_t>(1, segments.size()))
+    {
+        // move down names by one, q&d hack
+        segments[i - 1].name_id = segments[i].name_id;
+        segments[i].length = coordinate_calculation::greatCircleDistance(segments[i - 1].location,
+                                                                         segments[i].location);
+    }
+
+    float segment_length = 0.;
+    EdgeWeight segment_duration = 0;
+    std::size_t segment_start_index = 0;
+
+    double path_length = 0;
+
+    for (const auto i : osrm::irange<std::size_t>(1, segments.size()))
+    {
+        path_length += segments[i].length;
+        segment_length += segments[i].length;
+        segment_duration += segments[i].duration;
+        segments[segment_start_index].length = segment_length;
+        segments[segment_start_index].duration = segment_duration;
+
+        if (TurnInstruction::NoTurn != segments[i].turn_instruction)
+        {
+            BOOST_ASSERT(segments[i].necessary);
+            segment_length = 0;
+            segment_duration = 0;
+            segment_start_index = i;
+        }
+    }
+
+    total_distance = static_cast<std::uint32_t>(round(path_length));
+    total_duration = static_cast<std::uint32_t>(round(
+        (extract_alternative ? raw_route.alternative_path_length : raw_route.shortest_path_length) /
+        10.));
+
+    auto start_phantom = raw_route.segment_end_coordinates.front().source_phantom;
+    auto target_phantom = raw_route.segment_end_coordinates.back().target_phantom;
+
+    // Post-processing to remove empty or nearly empty path segments
+    if (segments.size() > 2 && std::numeric_limits<float>::epsilon() > segments.back().length &&
+        !(segments.end() - 2)->is_via_location)
+    {
+        segments.pop_back();
+        segments.back().necessary = true;
+        segments.back().turn_instruction = TurnInstruction::NoTurn;
+        target_phantom.name_id =
+            (segments.end() - 2)
+                ->name_id; // TODO check whether this -2 is desired after the pop-back
+    }
+
+    if (segments.size() > 2 && std::numeric_limits<float>::epsilon() > segments.front().length &&
+        !(segments.begin() + 1)->is_via_location)
+    {
+        segments.erase(segments.begin());
+        segments.front().turn_instruction = TurnInstruction::HeadOn;
+        segments.front().necessary = true;
+        start_phantom.name_id = segments.front().name_id;
+    }
+
+    if (allow_simplification)
+    {
+        DouglasPeucker polyline_generalizer;
```

Comment:
Can we cache and keep around a single `DouglasPeuker`? That is, making it static or storing outside.